### PR TITLE
Update date fields for unpaid work

### PR DESCRIPTION
**Pull request by raamSoftwire**
## Context

Edit the date fields for unpaid work to match those on the job form. This work is behind the restructured work history flag.

### Before
image

### After

image

## Guidance to review

Go commit by commit

## Link to Trello card

https://trello.com/c/8MQGKSwr/3063-unpaid-work-experience-update-date-fields-to-match-those-used-in-the-add-a-job-form

## Things to check

- [x] This code does not rely on migrations in the same Pull Request
- [x] If this code includes a migration adding or changing columns, it also backfills existing records for consistency
- [x] API release notes have been updated if necessary
- [x] New environment variables have been [added to the Azure config](https://github.com/DFE-Digital/apply-for-teacher-training/blob/master/docs/environment-variables.md#azure-hosting-devops-pipeline)

### CODE REVIEW

File: spec/forms/candidate_interface/volunteering_role_form_feature_flag_on_spec.rb [hunk 1]
```diff
@@ -0,0 +1,176 @@
+require 'rails_helper'
```
Comment (raamSoftwire) [inline spec/forms/candidate_interface/volunteering_role_form_feature_flag_on_spec.rb:1]: This is a straight copy from the existing form spec. No changes have been made in this commit.

File: app/forms/candidate_interface/volunteering_role_form.rb [hunk 2]
```diff
@@ -3,16 +3,38 @@ class VolunteeringRoleForm
...
+              :organisation,
+              :working_with_children,
+              presence: true
 
     validates :role, :organisation, length: { maximum: 60 }
-
     validates :start_date, date: { presence: true, future: true, month_and_year: true, before: :end_date }
-    validates :end_date, date: { future: true, month_and_year: true }, unless: :start_date_blank?
+
+    with_options if: :restructured_work_history_flag_active? do |form|
```
Comment (TobyRet) [inline app/forms/candidate_interface/volunteering_role_form.rb:29]: Nice. Was not aware you could group validations like this

Comment (raamSoftwire) [reply app/forms/candidate_interface/volunteering_role_form.rb:29]: Neither did I. But I spent a lot of time reading this page https://guides.rubyonrails.org/active_record_validations.html#grouping-conditional-validations

I think it improves the readability and makes it clear which context those validations should be executed in.

File: app/forms/candidate_interface/volunteering_role_form.rb [hunk 3]
```diff
@@ -3,16 +3,38 @@ class VolunteeringRoleForm
...
+      form.validates :start_date_unknown, inclusion: { in: %w[true false] }
+      form.validates :end_date_unknown, inclusion: { in: %w[true false] }
+      form.validates :currently_working, presence: true
+      form.validates :currently_working, inclusion: { in: %w[true false] }, if: -> { currently_working.present? }
+      form.validates :end_date, date: { presence: true, future: true, month_and_year: true }, if: -> { currently_working == 'false' }
```
Comment (TobyRet) [inline app/forms/candidate_interface/volunteering_role_form.rb:34]: `currently_volunteering` as opposed to `currently_working`?

File: spec/forms/candidate_interface/volunteering_role_form_feature_flag_on_spec.rb [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+
+RSpec.describe CandidateInterface::VolunteeringRoleForm, type: :model do
+  FeatureFlag.activate('restructured_work_history')
```
Comment (TobyRet) [inline spec/forms/candidate_interface/volunteering_role_form_feature_flag_on_spec.rb:4]: Wondering if this needs to be in a `before` block?
Would having it here cause state to leak out into other specs and cause flakiness?

File: spec/forms/candidate_interface/volunteering_role_form_spec.rb [hunk 5]
```diff
@@ -1,6 +1,8 @@
 require 'rails_helper'
 
 RSpec.describe CandidateInterface::VolunteeringRoleForm, type: :model do
+  FeatureFlag.deactivate('restructured_work_history')
```
Comment (TobyRet) [inline spec/forms/candidate_interface/volunteering_role_form_spec.rb:4]: https://github.com/DFE-Digital/apply-for-teacher-training/pull/4213/files#r592392701

File: spec/system/candidate_interface/entering_details/candidate_edits_or_deletes_volunteering_section_after_competing_the_section_spec.rb [hunk 6]
```diff
@@ -4,6 +4,8 @@
   include CandidateHelper
 
   scenario 'Candidate adds or deletes a role after completing the section' do
+    FeatureFlag.deactivate('restructured_work_history')
```
Comment (TobyRet) [inline spec/system/candidate_interface/entering_details/candidate_edits_or_deletes_volunteering_section_after_competing_the_section_spec.rb:7]: https://github.com/DFE-Digital/apply-for-teacher-training/pull/4213/files#r592392701

File: spec/system/candidate_interface/entering_details/candidate_edits_or_deletes_volunteering_section_with_restructured_work_history_active_spec.rb [hunk 7]
```diff
@@ -0,0 +1,105 @@
+require 'rails_helper'
+
+RSpec.feature 'Candidate edits their volunteering section' do
+  include CandidateHelper
+
+  scenario 'Candidate adds or deletes a role after completing the section' do
+    FeatureFlag.activate('restructured_work_history')
```
Comment (TobyRet) [inline spec/system/candidate_interface/entering_details/candidate_edits_or_deletes_volunteering_section_with_restructured_work_history_active_spec.rb:7]: https://github.com/DFE-Digital/apply-for-teacher-training/pull/4213/files#r592392701

... And wrap in a method `given_the_foo_is_bar`

File: spec/system/candidate_interface/entering_details/candidate_entering_volunteering_experience_spec.rb [hunk 8]
```diff
@@ -0,0 +1,104 @@
+require 'rails_helper'
+
+RSpec.feature 'Entering volunteering experience' do
+  include CandidateHelper
+
+  scenario 'Candidate adds volunteering experience' do
+    FeatureFlag.activate('restructured_work_history')
```
Comment (TobyRet) [inline spec/system/candidate_interface/entering_details/candidate_entering_volunteering_experience_spec.rb:7]: https://github.com/DFE-Digital/apply-for-teacher-training/pull/4213/files#r592394731

**Review by TobyRet — CHANGES_REQUESTED** [review]
Almost there. Some minor tweaks required!

File: app/forms/candidate_interface/volunteering_role_form.rb (hunk 2 above)
Comment (TobyRet) [reply app/forms/candidate_interface/volunteering_role_form.rb:29]: Yep absolutely

File: spec/system/candidate_interface/entering_details/candidate_edits_or_deletes_volunteering_section_with_restructured_work_history_active_spec.rb (hunk 7 above)
Comment (raamSoftwire) [reply spec/system/candidate_interface/entering_details/candidate_edits_or_deletes_volunteering_section_with_restructured_work_history_active_spec.rb:7]: I got rid of that change in a later commit

File: app/forms/candidate_interface/volunteering_role_form.rb (hunk 3 above)
Comment (raamSoftwire) [reply app/forms/candidate_interface/volunteering_role_form.rb:34]: Work experiences and volunteering experiences use the same DB table and this refers to a column `currently_working` so I think it's best we leave it as it is. 

The copy on the designs also says "Are you still working in this role?" for the view, so I think it's fine to stick with `currently_working` here too

Comment (TobyRet) [reply app/forms/candidate_interface/volunteering_role_form.rb:34]: 👍

**Review by TobyRet — APPROVED** [review]

**Pull request merged**